FABulous/timing_model/hdlnx/verilog_gate_level.py

```diff
@@ -0,0 +1,1041 @@
+        if not isinstance(self.debug, bool):
+            raise TypeError("debug must be a boolean.")
+    
+    def _call_external(self, executable: str, args: list[str] = [], 
+                       stdin_data: str = "", debug: bool = False) -> subprocess.CompletedProcess:
+        """
+        Calls an external executable with given arguments and stdin data.
+        Captures the output and checks for errors.
+        Args:
+            executable (str): The path to the executable to run.
+            args (list[str]): List of arguments to pass to the executable.
+            stdin_data (str): Data to send to the executable's stdin.
+        Returns:
+            subprocess.CompletedProcess: The result of the subprocess call.
+        Raises:
+            RuntimeError: If the external command fails.
+        """
+        
+        if debug:
+            print("Debug mode enabled for external command.")
+            print(f"Calling external command: {executable} {' '.join(args)}")
+            print(f"With stdin data:\n{stdin_data}")
+            result = subprocess.run(
+                [executable, *args],
+                input=stdin_data,   
+                text=True,                   
+            )
+        else:
+            result = subprocess.run(
+                [executable, *args],
+                input=stdin_data,   
+                text=True,            
+                capture_output=True, 
+                check=False,          
+            )
+        
+        if result.returncode != 0:
+            raise RuntimeError(f"Command '{' '.join([executable, *args])}' failed with error: {result.stderr}")
+        return result
+    
+    ### Protected methods (STA connection Interface) ###
+        
+    def _generate_sdf_opensta(self) -> Path:
+        """
+        Generates an temporary SDF file from the Verilog gate-level netlist using OpenSTA.
+        The SDF file is created in a temporary location and deleted after use.
+        Returns:
+            Path: The path to the generated SDF file.
+        """
+        
+        sta_tcl_script: str = ""
+        if isinstance(self.liberty_files, Path):
+            sta_tcl_script += f"read_liberty {self.liberty_files}\n"
+        else:
+            for lib in self.liberty_files:
+                sta_tcl_script += f"read_liberty {lib}\n"
+        sta_tcl_script += f"read_verilog {self.verilog_netlist}\n"
+        sta_tcl_script += f"link_design {self.top_name}\n"
+        if self.spef_files is not None:
+            if isinstance(self.spef_files, Path):
+                sta_tcl_script += f"read_spef {self.spef_files}\n"
+            elif isinstance(self.spef_files, list):
+                for spef in self.spef_files:
+                    sta_tcl_script += f"read_spef {spef}\n"
+        sta_tcl_script += "write_sdf {}\n".format("{sdf_path}")
+        sta_tcl_script += "exit\n"
+        
+        fd, path = tempfile.mkstemp(prefix="sta_", suffix=".sdf")
```

Comment:
If the file is intended to be temporary, you can use `with` to manage the context.